### PR TITLE
HDDS-4594. Add namespace quota to documentation

### PR DESCRIPTION
**Pull request by amaliujia**
## What changes were proposed in this pull request?

Add namespace quota to documentation

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4594

## How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @captainzmc

File: hadoop-hdds/docs/content/feature/Quota.md [hunk 1]
```diff
@@ -45,6 +45,18 @@ e. If the cluster is upgraded from old version less than 1.1.0, use of quota on
...
+Administrators should be able to define how many namespace a Volume or Bucket can use. The following settings for namespace quota are supported: 
+
+a. By default, the namespace quota for volume and bucket is not enabled.
+
+b. When volume namespace quota is enabled, the total number of buckets under the volume, cannot exceed the volume namespace quota.
+
+c. When bucket namespace quota is enabled, the totoal number of keys under the bucket, cannot exceed the bucket namespace quota.
+
+d. Linked buckets do not consume namespace quota.
+
```
Comment (amaliujia) [inline hadoop-hdds/docs/content/feature/Quota.md:59]: Do we also need example client usage for namespace quota in the following section?

File: hadoop-hdds/docs/content/feature/Quota.md [hunk 2]
```diff
@@ -45,6 +45,18 @@ e. If the cluster is upgraded from old version less than 1.1.0, use of quota on
...
+Administrators should be able to define how many namespace a Volume or Bucket can use. The following settings for namespace quota are supported: 
+
+a. By default, the namespace quota for volume and bucket is not enabled.
+
+b. When volume namespace quota is enabled, the total number of buckets under the volume, cannot exceed the volume namespace quota.
+
+c. When bucket namespace quota is enabled, the totoal number of keys under the bucket, cannot exceed the bucket namespace quota.
+
+d. Linked buckets do not consume namespace quota.
```
Comment (captainzmc) [inline hadoop-hdds/docs/content/feature/Quota.md:58]: Whether 'e' in Storage Space applies equally to Namespace quota. If so, we need to list.

**Comment by captainzmc** [comment]
@amaliujia Thanks for your patch. Could you also update the Client Usage about Namespace quota?

**Comment by amaliujia** [comment]
@captainzmc 

comments addressed.

**Review by linyiqun — COMMENTED** [review]
When reading this quota documentation, I am thinking for the set quota cases: as we support both volume and bucket quota, will we check the below scenarios that only either one quota was set, then try to enable another quota:

* Bucket quota already set, then user try to set volume quota, will we check volume quota must more than total bucket quota?
* Volume quota already set, then user try to set bucket quota, will we check the total bucket quota, so that will not exceed volume quota? (I remember this check already contained)

**Comment by amaliujia** [comment]
@linyiqun 

I think for namespace, it will be different than space. Namespace quota is independent on each level (volume, bucket, etc.)

Namespace quota on volume only controls number of bucket names, and namespace on bucket only controls keys in that bucket. This is different from space quota in which volume quota is a relationship with its buckets quotas.

**Comment by captainzmc** [comment]
Thanks for @amaliujia's update. I think there are some case we need to confirm.
Whether 'f' in Storage Space applies equally to Namespace quota?  If so, we need to list.  This part of namespace has been modified in #1723 (reviewing).

File: hadoop-hdds/docs/content/feature/Quota.md [hunk 3]
```diff
@@ -73,10 +87,35 @@ This behavior changes the quota for Bucket1 to 10GB
...
+bin/ozone sh bucket setquota --namespace-quota 1000 /volume1/bucket1 
+```
+This behavior changes the quota for Bucket1 to 1000.
+
 #### Clear the quota for volume and bucket
 ```shell
-bin/ozone sh volume clrquota --space-quota /volume1
-bin/ozone sh bucket clrquota --space-quota /volume1/bucket1
+bin/ozone sh volume clrquota --space-quota --namespace-quota /volume1
+bin/ozone sh bucket clrquota --space-quota --namespace-quota /volume1/bucket1
```
Comment (captainzmc) [inline hadoop-hdds/docs/content/feature/Quota.md:118]: **There's something wrong with the directory structure. At present is:**

#space quota
……
#namespace quota
……
## clear quota
## check quota and usage

**We can change it to:**

#space quota
……
## clear quota
## check quota and usage

#namespace quota
……
## clear quota
## check quota and usage

**or**

#space quota and namespace quota

……
## clear quota
## check quota and usage

File: hadoop-hdds/docs/content/feature/Quota.zh.md [hunk 4]
```diff
@@ -70,10 +84,35 @@ bin/ozone sh bucket setquota  --space-quota 10GB /volume1/bucket1
...
+bin/ozone sh bucket setquota --namespace-quota 1000 /volume1/bucket1 
+```
+该行为将bucket1的命名空间配额更改为1000。
+
 #### 清除volume和bucket的配额
 ```shell
-bin/ozone sh volume clrquota --space-quota /volume1
-bin/ozone sh bucket clrquota --space-quota /volume1/bucket1
+bin/ozone sh volume clrquota --space-quota --namespace-quota /volume1
+bin/ozone sh bucket clrquota --space-quota --namespace-quota /volume1/bucket1
```
Comment (captainzmc) [inline hadoop-hdds/docs/content/feature/Quota.zh.md:115]: There's something wrong with the directory structure. At present is:
`#space quota`
……
`#namespace quota`
……
`## clear quota`
`## check quota and usage`


We can change it to:

`#space quota`
……
`## clear quota`
`## check quota and usage`

`#namespace quota`
……
`## clear quota`
`## check quota and usage`

or

`#space quota and namespace quota`

……
`## clear quota`
`## check quota and usage`

**Comment by amaliujia** [comment]
@captainzmc 
f. does not apply to namespace in my opinion. Note that the difference so far for namespace quota is that it is set independently on volume and bucket (not like space quota).

**Comment by captainzmc** [comment]
> @captainzmc
> f. does not apply to namespace in my opinion. Note that the difference so far for namespace quota is that it is set independently on volume and bucket (not like space quota).

Agreed, I'll remove the namespace restriction in #1723.

**Comment by amaliujia** [comment]
@captainzmc 

finally I was able to re-trigger the CI (seemed that CI didn't work for a while). Any other comments on this PR?

**Review by captainzmc — APPROVED** [review]
LGTM +1
hi @linyiqun, any other suggestions about this PR?

**Comment by linyiqun** [comment]
Also +1 from me.

**Review by linyiqun — APPROVED** [review]

**Comment by captainzmc** [comment]
Merged this PR. Thanks for @amaliujia's patch and @linyiqun‘s review.

**Pull request merged**